Add predicted database search

pdmnyberg opened the pull request:
This PR closes #29 

It will:
- Clean up search code
- Clean up html
- Clean up typing
- Use statements instead of queries to simplify reuse
- Add pagination
- Add support for searching in the predicted database

To test:
- got to: http://localhost:5000/search
- use the search fields for both databases
- expect results

pdmnyberg committed:
fix: use placeholders instead of default values for peptide sequence length

pdmnyberg committed:
refactor: improve typing

pdmnyberg committed:
feat: add initial predicted database search

pdmnyberg committed:
fix: improve use of hr

pdmnyberg committed:
fix: replicated pagination at bottom of list

pdmnyberg committed:
chore: clean lint errors

pdmnyberg committed:
refactor: clean up types and make chemical class search more strict

pdmnyberg committed:
fix: only show pagination if available

pdmnyberg committed:
fix: add basic fixed ordering to advanced search

pdmnyberg committed:
fix: add prefetching of compounds

andkaha commented on app/search.py:
```diff
@@ -1,68 +1,137 @@
+        if chemical_class_type == "class":
+            stmt = stmt.filter(
+                Validated.compounds.any(
+                    Compounds.chemical_class.ilike(f"%{chemical_class_name}%")
```
I don't think this should be a `LIKE` pattern match, but an identity comparison. The string in `chemical_class_name` is picked from a dropdown and is not entered as freetext, hence we know that the exact string should be available in the `compound.chemical_class` field. 

At the moment, if we select `class: Acid` from the dropdown, we get some results that contain no mention of `class: Acid` but of `class: Free Fatty Acids`.

This would also make it possible to add an index on that field, potentially speeding up this query.

pdmnyberg replied:
I see. Thanks!

pdmnyberg replied:
I am thinking that a case insensitive match would still be preferred in order to make sure we don't miss possible case variants. If `chemical_class` was normalized in the database I would trust it a bit more. Would `Compounds.chemical_class.ilike(chemical_class_name)` be good enough for you here or do you have a better suggeston?

pdmnyberg replied:
You know what. I withdraw my last response. Since the drop down is based on this field it would cause duplicates if we had case variants. So I will just make an exact match here. Thanks again

andkaha commented on app/search.py:
```diff
@@ -1,68 +1,137 @@
+            stmt = stmt.filter(
                 Validated.compounds.any(
-                    Compounds.chemical_class.ilike(f"%{chemical_class}%")
+                    Compounds.compound_name.ilike(f"%{chemical_class_name}%")
```
The same goes here; it should likely be an identity comparison, not a `LIKE` pattern match.

andkaha reviewed (changes requested):
A couple of changes requested, having to do with converting `LIKE` matching into identity comparisons.

pdmnyberg committed:
fix: adjust off by one for last page

pdmnyberg committed:
fix: use exact matches for chemical class and compound name

pdmnyberg committed:
fix: add page info to status

andkaha committed:
Improve indexing

* Add non-unique index on compounds.chemical_class.
* Swap validated_id and compound_id in the unique index in
  validated_compounds to improve query planning when filtering using a
  compound.
* Add non-unique index on predicted_unique_homologues.validated_id to
  improve selecting predicted homologues given data from the validated
  table.

andkaha committed:
Merge branch '29-add-predicted-database-search' of github.com:NBISweden/bacmet into 29-add-predicted-database-search

andkaha reviewed (approved):
Looks good now.
